### PR TITLE
Undo will now remove the project or metadata references added by add-import

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/10178

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
retest windows_release_unit32_prtest  please

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs [hunk 1]
```diff
@@ -191,84 +191,88 @@ public ITextBufferAssociatedViewService AssociatedViewService
...
-                    if (changedFiles.Count == 0)
-                    {
-                        operation.Apply(workspace, cancellationToken);
-                        continue;
-                    }
+            var applyOperation = operationsList.Single() as ApplyChangesOperation;
+            if (applyOperation == null)
+            {
+                return null;
+            }
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: It seems uber-paranoid to test `Length` and then use `Single()`. Couldn't you just have the following and remove the code that checks the `Length > 1` above?

```C#
var applyOperation = operationList.SingleOrDefault() as ApplyChangesOperation;
if (applyOperation == null)
{
    return null;
}
```

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs [hunk 2]
```diff
@@ -191,84 +191,88 @@ public ITextBufferAssociatedViewService AssociatedViewService
...
 
-                        // link current file in the global undo transaction
-                        if (fromDocument != null)
-                        {
-                            undoTransaction.AddDocument(fromDocument.Id);
-                        }
+            var projectChange = projectChanges.Single();
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: If you use `SingleOrDefault()`, you could avoid the allocation of an `ImmutableArray<T>` and iterating the whole `IEnumerable<ProjectChanges>`.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.AddProjectReferenceUndoUnit.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                    var updatedProject = fromProject.AddProjectReference(new ProjectReference(_toProjectId));
+                    Workspace.TryApplyChanges(updatedProject.Solution);
+                }
+            }
+
+            public override void GetDescription(out string pBstr)
+            {
+                var currentSolution = Workspace.CurrentSolution;
+                var toProject = currentSolution.GetProject(_toProjectId);
+                var projectName = toProject?.Name ?? "";
```
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.AddProjectReferenceUndoUnit.cs]: `string.Empty`?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 4]
```diff
@@ -420,9 +430,33 @@ protected override void ApplyProjectReferenceAdded(ProjectId projectId, ProjectR
...
+            undoManager?.Add(new RemoveProjectReferenceUndoUnit(
+                this, projectId, projectReference.ProjectId));
+        }
+
+        private OleInterop.IOleUndoManager TryGetUndoManager()
+        {
+            var documentTrackingService = this.Services.GetService<IDocumentTrackingService>();
+            if (documentTrackingService != null)
+            {
+                var documentId = documentTrackingService.GetActiveDocument() ?? documentTrackingService.GetVisibleDocuments().FirstOrDefault();
```
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Seems like a nice candidate for a helper. Do we do this elsewhere in the code base?

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: SingleOrDefault throws if you have more than 1.  http://index/#System.Core/System/Linq/Enumerable.cs,1087

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: It does not appear so, no.

File: src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs]: I'd forgotten. Always throws me for a loop (ha!)

**Pull request merged**